File: annotation-processors/src/test/java/com/comsysto/livingdoc/annotation/processors/plantuml/PlantUmlClassDiagramProcessorTest.java
package com.comsysto.livingdoc.annotation.processors.plantuml;

import static com.comsysto.livingdoc.annotation.processors.plantuml.PlantUmlClassDiagramProcessor.DEF_OUT_DIR;
import static com.comsysto.livingdoc.annotation.processors.plantuml.PlantUmlClassDiagramProcessor.KEY_OUT_DIR;
import static java.util.stream.Collectors.toList;
import static org.assertj.core.api.Assertions.assertThat;

import com.google.testing.compile.Compilation;
import com.google.testing.compile.Compilation.Status;
import com.google.testing.compile.Compiler;
import lombok.SneakyThrows;

import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.nio.file.FileSystemNotFoundException;
import java.util.Arrays;
import java.util.Optional;

@SuppressWarnings("ConstantConditions")
public class PlantUmlClassDiagramProcessorTest {

    private File exampleDir;

    @Before
    public void setUp() {
        final File testRoot = Optional.ofNullable(System.getProperty(PlantUmlDiagramTestUtils.TEST_DIR))
            .map(File::new)
            .orElseGet(() -> PlantUmlDiagramTestUtils.getTestClassLoaderRoot(this.getClass().getClassLoader()));
        exampleDir = new File(testRoot, "example");
    }

    @Test
    @SneakyThrows
    public void should_run_successfully_and_produce_expected_output() {
        final Compilation result = Compiler.javac()
            .withProcessors(new PlantUmlClassDiagramProcessor())
            .withOptions("-Apumlgen.out.dir=out")
            .compile(Optional.ofNullable(exampleDir.listFiles())
                         .map(Arrays::stream)
                         .orElseThrow(() -> new FileSystemNotFoundException(exampleDir.getAbsolutePath()))
                         .map(PlantUmlDiagramTestUtils::javaFileObject)
                         .collect(toList()));

        assertThat(result.status()).isEqualTo(Status.SUCCESS);

        // Check that the file exists and contains a selected sample of expected
        // lines:
        assertThat(new File(System.getProperty(KEY_OUT_DIR, DEF_OUT_DIR), "package_class.puml"))
            .exists()
            .satisfies(file -> assertThat(PlantUmlDiagramTestUtils.lines(file))
                .contains(
                    "class Car",
                    "Wing leftWing",
                    "Flying <|.. FlyingVehicle",
                    "FlyingVehicle <|-- Airplane",
                    "Airplane --> Wing"));
    }

}

File: annotation-processors-kotlin/src/test/kotlin/annotation/processors/plantuml/PlantUmlClassDiagramProcessorTest.kt
package annotation.processors.plantuml

import org.apache.log4j.BasicConfigurator
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths
import java.util.Arrays.stream
import java.util.stream.Collectors.toSet
import javax.tools.JavaFileObject

class PlantUmlClassDiagramProcessorTest {
    private var exampleDir: File? = null

    private val testClassLoaderRoot: File
        get() {
            val projectRoot = Paths.get(this.javaClass.getClassLoader().getResource(".")!!.toURI())
                    .getParent()
                    .getParent()
                    .getParent()
                    .toFile()
            return File(projectRoot, "src/test/java")
        }

    @Before
    fun setUp() {
        val testRoot = File(System.getProperty(TEST_DIR)) ?: getTestClassLoaderRoot()
        exampleDir = File(testRoot, "com/comsysto/livingdoc/example")
    }

    @Test
    fun should_run_successfully_and_produce_expected_output() {
        val result = Compiler.javac()
                .withProcessors(PlantUmlClassDiagramProcessor())
                .withOptions("-Apumlgen.out.dir=out")
                .compile(stream(exampleDir!!.listFiles()!!)
                        .map { this.javaFileObject(it) })

        assertThat(result.status()).isEqualTo(Status.SUCCESS)

        // Check that the file exists and contains a selected sample of expected
        // lines:
        assertThat(File(System.getProperty(KEY_OUT_DIR, DEF_OUT_DIR), "package_class.puml"))
                .exists()
                .satisfies({ file ->
                    assertThat<String>(lines(file))
                            .contains(
                                    "class Car",
                                    "Wing leftWing",
                                    "Flying <|.. FlyingVehicle",
                                    "FlyingVehicle <|-- Airplane",
                                    "Airplane --> Wing")
                })
    }


    private fun lines(file: File): Set<String> {
        return Files.lines(file.toPath())
                .map { it.trim({ it <= ' ' }) })
    }


    private fun javaFileObject(f: File): JavaFileObject {
        return JavaFileObjects.forResource(f.toURI().toURL())
    }

    companion object {

        protected val TEST_DIR = "pumlgen.test.dir"

        @BeforeClass
        fun setupBeforeClass() {
            BasicConfigurator.configure()
        }
    }
}